core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -438,9 +443,10 @@ public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) t
       CompletableFuture<Void> initialStage = new CompletableFuture<>();
       CompletionStage<Void> currentStage = initialStage;
       for (Object key : command.getMap().keySet()) {
+         int segment = keyPartitioner.getSegment(key);
          // as listeners may need the value, we'll load the previous value
-         currentStage = entryFactory.wrapEntryForWriting(ctx, key, keyPartitioner.getSegment(key),
-               ignoreOwnership || canReadKey(key), command.loadType() != VisitableCommand.LoadType.DONT_LOAD, currentStage);
+         currentStage = entryFactory.wrapEntryForWriting(ctx, key, segment,
+               ignoreOwnership || canReadSegment(segment) || needsWriteKeyForSegment(segment), command.loadType() != VisitableCommand.LoadType.DONT_LOAD, currentStage);
```

Comment:
This is the actual fix as we now always wrap the entry in the context as an MVCCEntry which is required by the CallInterceptor for the PutAllCommand.